state/src/impls/top_level.rs

```diff
@@ -481,10 +537,15 @@ impl TopLevelState {
         // FIXME: Make it mutable borrow db instead of cloning.
         let mut shard_level_state = ShardLevelState::from_existing(shard_id, self.db.clone(), shard_root)?;
 
-        let invoice = shard_level_state.apply(transaction, sender, &shard_users, client)?;
-        let (new_root, db) = shard_level_state.drop();
+        let invoice = shard_level_state.apply(&transaction.clone().into(), sender, &shard_users, client)?;
         match &invoice {
             Invoice::Success => {
+                let unwrapped_amount = transaction.unwrapped_amount();
+                if !unwrapped_amount.is_zero() {
+                    self.add_balance(sender, &unwrapped_amount)?;
```

Comment:
It must be after `self.db = db;` statement.

Reply:
https://github.com/CodeChain-io/codechain/pull/854 will make it share `db` and it will fix this constraint. But the current implementation replaces `db`, so changing top level state must be done after finish the shard level state changing.

Reply:
Thanks, I will fix this.

Reply:
Sorry. It must be moved to `apply_transaction` method. The current position executed several time if the inputs of the `UnwrapCCC` transaction are in multiple shards.

Reply:
Yes, that was what I was concerned about, but I thought it is okay because `transaction.related_shards()` only return one shard on `AssetUnwrapCCC`.

Reply:
But it's not proper to put in there by its meaning. I'll fix it